File: Programmers/여행경로.java
import java.util.*;

class Solution {
    static List<String> answer = new ArrayList<>();
    static String route = "";
    static boolean[] visited;
    
    public String[] solution(String[][] tickets) {
        visited = new boolean[tickets.length];
        dfs(tickets, "ICN", 0);

        Collections.sort(answer);
        return answer.get(0).split(",");
    }
    
    public static void dfs(String[][] tickets, String end, int cnt) {
        route += end + ",";

        if (cnt == tickets.length) {
            answer.add(route);
            return;
        }

        for (int i = 0; i < tickets.length; i++) {
            String next_s = tickets[i][0];
            String next_e = tickets[i][1];
            if (next_s.equals(end) && !visited[i]) {
                visited[i] = true;
                dfs(tickets, next_e, cnt + 1);
                visited[i] = false;
                route = route.substring(0, route.length() - 4);
            }
        }
    }
}

File: Programmers_python/여행경로.py
from collections import deque

answer = []

def dfs(tickets, visited, cur, result):
    global answer
        
    if len(list(result.split())) == len(tickets) + 1:
        answer.append(result)
        return
    
    for i in range(len(tickets)):
        if not visited[i] and tickets[i][0] == cur[1]:
            visited[i] = True
            dfs(tickets, visited, tickets[i], result+" "+tickets[i][1])
            visited[i] = False
    return
    
def solution(tickets):
    visited = [False] * len(tickets)
    tickets.sort()
    
    for i in range(len(tickets)):
        if tickets[i][0] == 'ICN':
            visited[i] = True
            dfs(tickets, visited, tickets[i], tickets[i][0]+" "+tickets[i][1])
            visited[i] = False
            
    if len(answer) >= 2:
        answer.sort()
    
    return list(answer[0].split())
